text/0000-write-pointers.md

```diff
@@ -0,0 +1,463 @@
+
+- RFC PR: (leave this empty)
+
+- Rust Issue: (leave this empty)
+
+# Summary
+
+[summary]: #summary
+
+This RFC aims to allow direct initialization for optimization, and partial struct and enum initialization, for ergonomics. It will do so through the usage of a two new reference type, `&out T` and `&uninit T` (the name is not important to me).
+
+# Motivation
+
+[motivation]: #motivation
+
+The builder pattern was created as a way to try and solve the issue of not having partial initialization, but it has problems with large structs, and that the `*Builder` struct must necessarily be larger than the target struct, null-ptr optimizations not-withstanding. Also, it is very expensive to move large structs, and relying on the optimizer to optimize out moves isn't very good, `&out T` could serve as a way to directly place things into the desired memory location and to maintain write-only memory. `&uninit T` will serve the purpose of partial initialization and direct initialization.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+`&out T` is a write-only reference to `T`, where `T: Copy`. The bound is necessary as it is not safe to overwrite `Drop` types without invoking the destructor. Does not run destructors on write. See [unresolved questions](#unresolved-questions) for a more in depth explanation. \
+`&uninit T` is a write-once reference to `T`, after the first write, it is allowed to read the values (behaves exactly like `&mut T`). Does not run destructors on the first write. \
+
+For all examples, I will use these two structs
+
+```Rust
+struct Foo { a: u32, b: String, c: Bar }
+#[derive(Clone, Copy)]
+struct Bar { d: u32, e: f32 }
+
+impl std::ops::Drop for Foo {
+    fn drop(&mut T) {
+        println!("Dropping Foo {}", foo.b);
+    }
+}
+```
+
+## `&uninit T`
+
+Using `&uninit T`, we can do partial initialization and directly initialize.
+```Rust
+let x: Foo;
+
+*(&uninit x.a) = 12;
+*(&uninit x.b) = "Hello World".to_string();
+*(&uninit x.c.d) = 11;
+*(&uninit x.c.e) = 10.0;
+```
+This works because when we take an `&uninit` to `x.a`, we are implicity also taking an `&uninit` to `x`, and the dot operator will not attempt to read the memory location anywhere in `x`.
+
+For ease of use, you can simply write
+```Rust
+let x: Foo;
+
+x.a = 12;
+x.b = "Hello World".to_string();
+x.c.d = 11;
+x.c.e = 10.0;
+```
+and the compiler will infer that all of these need to use `&uninit`, because `x` was not initialized directly.
+
+### Restrictions
+
+#### Storing
+
+You cannot store `&uninit T` in any way, not in structs, enums, unions, or behind any references. So all of these are invalid.
+
+```Rust
+fn maybe_init(maybe: Option<&uninit T>) { ... }
+fn init(ref_to_write: &mut &uninit T) { ... }
+struct Temp { a: &uninit Foo }
+```
+
+#### Conditional Initialization
+
+One restriction to `&uninit T` is that we cannot conditionally initialize a value. For example, none of these are allowed.
```

Comment:
`&uninit T` is more like a compiler hint than an actual type, and storing doesn't make much sense. Also, in regards to generic types.

```Rust
let x: u32;
Option::Some::<&uninit u32>(&uninit x);
```
This would be invalid, `Some` is a function that takes a `&uninit`, but doesn't initialize it.

Also in general, generics cannot handle `&uninit T`, with the current restrictions on `&uninit T`, namely that if a function takes an `&uninit T`, then it must be initialized in that function.

Reply:
Because there a different behavior on the first write for `&uninit T`, it does not attempt to drop the old value, because there is no old value. I will add this to unresolved questions because it seems like this is feasible.